### DISCUSSION TITLE
socks5 proxy

### DISCUSSION THREAD

**Discussion by Masber** [kube-rs/kube#1669] [Q&A]
Hi, could you please share an example using kube-rs to create a kubernetes client using socks5 proxy?

**Comment by clux** [comment]
It should just work out of the box with `Client::try_default()` provided you have the `socks5` [feature](https://kube.rs/features/) and a cluster configured using `proxy-url` in the kubeconfig. 

If you need a custom client, try using the [custom_client example](https://github.com/kube-rs/kube/blob/main/examples/custom_client.rs) but use a more specific `connector` rather than the default the example is using;

```rust
let mut base_connector = HttpConnector::new();
base_connector.enforce_http(false);
let proxy_connector = hyper_socks2::SocksConnector {
    proxy_addr: proxy_url.clone(),
    auth: None,
    base_connector,
};
let connector = config.rustls_https_connector_with_connector(proxy_connector)?;
```

See our [internal builder setup](https://github.com/kube-rs/kube/blob/0424cb48911b63c1e1572f2f382d9f3119c9f271/kube-client/src/client/builder.rs#L94-L111) for some more ideas there.

## Example `ssh` Setup

A setup thing I usually test with is to socks proxy to another machine over `ssh`.

If you have a `k3sbox`, then grab its kubeconfig off that server, add `proxy-url: socks5://0.0.0.0:3126` to `clusters.cluster[0]` and use that kubeconfig on your main machine.

add an entry in `~/.ssh/config` for forwarding:

```aconf
Host k3sbox
  HostName mymachinerunningk3s
  Port 8702
  User you
  IdentityFile ~/.ssh/k3s_id
  DynamicForward 3126
```

then keep a terminal maintaining an `ssh k3sbox` connection, and any `kubectl` or `kube::Client` stuff will then use the `proxy-url` to bounce through 0.0.0.0:3126 which will ssh forward through k3sbox and then use the regular server url on the other side.
